### ISSUE TITLE
FAPI integration tests hang at 100% CPU usage after enabling async mode for tcti-mssim 

### ISSUE THREAD

**Issue by diabonas** [tpm2-software/tpm2-tss#1822]
Since commit 6a7d845ab241ecb936875728e908298e567aa203, individual FAPI integration tests randomly hang sometimes and consume a lot of CPU when using the mssim TCTI:
```
./bootstrap
./configure --disable-tcti-swtpm --enable-integration
make check TESTS=test/integration/fapi-get-random
```
A backtrace of a stuck integration test looks as follows:
```
#0  0x00007f6673872027 in poll () from /usr/lib/libc.so.6
#1  0x00007f6673f17652 in tcti_mssim_receive (tctiContext=0x561020b81400, response_size=0x7ffd7a6302c8, response_buffer=0x561020b82a70 "", timeout=0)
    at src/tss2-tcti/tcti-mssim.c:403
#2  0x00007f6673f34caa in tctildr_receive (tctiContext=0x561020b81480, response_size=0x7ffd7a6302c8, response_buffer=0x561020b82a70 "", timeout=0)
    at src/tss2-tcti/tctildr.c:206
#3  0x00007f6673edecff in Tss2_Sys_ExecuteFinish (sysContext=0x561020b82a20, timeout=0) at src/tss2-sys/api/Tss2_Sys_Execute.c:82
#4  0x00007f6673e3ef80 in Esys_GetCapability_Finish (esysContext=0x561020b814f0, moreData=0x7ffd7a63034b "", capabilityData=0x561020b757d0)
    at src/tss2-esys/api/Esys_GetCapability.c:273
#5  0x00007f6673d2d663 in Fapi_Initialize_Finish (context=0x56101eed3258 <global_fapi_context>) at src/tss2-fapi/api/Fapi_Initialize.c:251
#6  0x00007f6673d2ca1c in Fapi_Initialize (context=0x56101eed3258 <global_fapi_context>, uri=0x0) at src/tss2-fapi/api/Fapi_Initialize.c:81
#7  0x000056101eecff0a in init_fapi (profile=0x56101eed1389 "P_ECC", fapi_context=0x56101eed3258 <global_fapi_context>) at test/integration/main-fapi.c:249
#8  0x000056101eed0140 in main (argc=1, argv=0x7ffd7a631298) at test/integration/main-fapi.c:294
```
This issue [can be reproduced on Travis CI](https://travis-ci.org/github/tpm2-software/tpm2-tss/jobs/715161354) as well after fixing a typo that was preventing the test suite from getting run with mssim at all, see #1821. The exact tests that get stuck vary from run to run, so I suspect some sort of race condition.

**Comment by tstruk** [comment]
It should be fixed by this commit:
https://github.com/tpm2-software/tpm2-tss/commit/e66284a6f434618bafd5d5bd81e1a68ab0433ab9
@diabonas Could you verify if it fixes it for you.

**Comment by diabonas** [comment]
Sorry, I forgot to mention which state of the repository I was using to test this: it was the current master branch, commit 1361aa370994eea08b446f1de975f918f9574e8c. Since commit e66284a6f434618bafd5d5bd81e1a68ab0433ab9 was already merged at that time, it does not seem to fully fix the issue.

**Comment by tstruk** [comment]
I will look into it now. The commit https://github.com/tpm2-software/tpm2-tss/pull/1793/commits/e6ed375b4d532801e7a80e5bde6a25956136f80e also causes appveyor to fail.
It worked for me locally with configure --disable-tcti-swtpm, but it was some time ago.
If it will cause too much trouble I will revert it for now.

**Comment by tstruk** [comment]
From the other hand - it works in cirrus. We only have mssim there.

**Comment by diabonas** [comment]
Yes, I can consistently reproduce this with the current master 4b7441bbcf760400fd846e090f984586d319d4ae using the same compilation sequence like you, my configure output also matches yours. It's not a huge number of tests that don't complete each run, about 2 to 5 usually, but the sheer number of FAPI tests means I can never actually complete the whole test suite since a random test is bound to hang.

I guess it depends a bit on the CPU power and load, for Travis CI it failed [at least once](https://travis-ci.org/github/tpm2-software/tpm2-tss/jobs/715161354) (so it doesn't appear only to be a problem with my machine or configuration), but certainly seems to pass more often than not. I have yet to see Cirrus failing, but I don't know enough about FreeBSD to judge whether it could potentially be affected by this as well or not.

**Comment by tstruk** [comment]
Could you try this https://github.com/tpm2-software/tpm2-tss/pull/1824 and see if it helps?

**Comment by diabonas** [comment]
That seems to do it, I was able to run the full integration test suite five times in a row without any issues, thank you very much!

**Comment by diabonas** [comment]
Fixed by #1824.
